src/ORM/ResultSetFactory.php

```diff
@@ -0,0 +1,286 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         5.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\ORM;
+
+use Cake\Collection\Collection;
+use Cake\Datasource\EntityInterface;
+
+/**
+ * Factory class for generation ResulSet instances.
+ *
+ * It is responsible for correctly nesting result keys reported from the query
+ * and hydrating entities.
+ */
+class ResultSetFactory
+{
+    /**
+     * Default table instance
+     *
+     * @var \Cake\ORM\Table
+     */
+    protected Table $_defaultTable;
+
+    /**
+     * The default table alias
+     *
+     * @var string
+     */
+    protected string $_defaultAlias;
+
+    /**
+     * List of associations that should be placed under the `_matchingData`
+     * result key.
+     *
+     * @var array
+     */
+    protected array $_matchingMap = [];
+
+    /**
+     * List of associations that should be eager loaded.
+     *
+     * @var array
+     */
+    protected array $_containMap = [];
+
+    /**
+     * Map of fields that are fetched with their type and the table they belong to.
+     *
+     * @var array
+     */
+    protected array $_map = [];
+
+    /**
+     * List of matching associations and the column keys to expect
+     * from each of them.
+     *
+     * @var array
+     */
+    protected array $_matchingMapColumns = [];
+
+    /**
+     * Whether to hydrate results into entities.
+     *
+     * @var bool
+     */
+    protected bool $_hydrate = true;
+
+    /**
+     * Tracks value of $_autoFields property of $query passed to constructor.
+     *
+     * @var bool|null
+     */
+    protected ?bool $_autoFields = null;
+
+    /**
+     * The fully namespaced name of the class to use for hydrating results.
+     *
+     * @var string
+     */
+    protected string $_entityClass;
+
+    /**
+     * Constructor
+     *
+     * @param \Cake\ORM\Query $query Query from where results came.
+     * @param array $results Results array.
+     */
+    public function createResultSet(Query $query, array $results): ResultSet
+    {
+        $this->_defaultTable = $query->getRepository();
+        $this->_hydrate = $query->isHydrationEnabled();
+        $this->_autoFields = $query->isAutoFieldsEnabled();
+
+        $this->_entityClass = $this->_defaultTable->getEntityClass();
+        $this->_defaultAlias = $this->_defaultTable->getAlias();
```

Comment:
Should this factory have state? Seems like we could make all these properties into an intermediate object or array that is passed around to each step of result generation. That would let us use a bit less memory when doing ORM queries, as the intermediate object or array could be garbage collected.

Reply:
Okay, I'll update accordingly.